### PR TITLE
[stable/consul] Provide advanced configuration options for consul

### PR DESCRIPTION
**Pull request by mitom**
As per #1198 this adds another option to configure consul, allowing to pass in configuration by both `configMap`s and `secret`s. It would be fairly trivial to extend this to any other type that kubernetes can mount as a volume (like hostpath) if the list grows, or if there is a usecase.

It also allows to re-use parts of the configuration in case of multiple clusters as these can be separated into multiple objects.

There is one drawback, the new `ConsulConfig` option can't be set using helm's `--set`. The reason being `--set` can't set list of hashes at the moment (AFAIK). The suggested approach is to just use a hash, however in this scenario it could be possible to have a `secret` and `configMap` named the same and therefore a hash wouldn't work. The attribute could be separated into 2, which would solve the `--set` issue as well, however this would degrade functionality as the order in which configuration is loaded could not be influenced per type, not globally (without post-processing the files on the container).

There are no breaking changes or degradations.

### CODE REVIEW

**Comment by lachie83** [comment]
@mitom Thanks for the contrib. I'll review as soon as I can. If anyone else is available to review that would be great too. Thanks

**Comment by mitom** [comment]
A thing I discovered is that due to the fact that these are mounted as volumes and kubernetes won't allow the volume specification to change for pods in a statefulset after creation, the list can only be updated by re-creating the statefulset. This would lead to #1143 as all containers would attempt to leave the cluster and would also cause downtime.

The contents of each item in the list can be changed easily and the pods reloaded/roll-restarted.

File: stable/consul/README.md [hunk 1]
```diff
@@ -32,6 +32,8 @@ The following tables lists the configurable parameters of the consul chart and t
 | `ImagePullPolicy`       | Container pull policy                 | `Always`                                                   |
 | `Replicas`              | k8s statefulset replicas              | `3`                                                        |
 | `Component`             | k8s selector key                      | `consul`                                                   |
+| `ConsulConfig.[].type`  | either `secret` or `configMap`        | -                                                      |
```
Comment (unguiculus) [inline stable/consul/README.md]: I don't like the syntax you used. This is confusing. May just add `ConsulConfig` and add an example to `values.yaml`.

File: stable/consul/values.yaml [hunk 2]
```diff
@@ -45,6 +45,8 @@ uiService:
   enabled: true
   type: "NodePort"
 
+ConsulConfig: []
+
```
Comment (unguiculus) [inline stable/consul/values.yaml]: Add example.

```yaml
ConsulConfig: []
#  - type: secret
#    name: consul-defaults
#  - type: configMap
#    name: consul-defaults
```

**Review by unguiculus — DISMISSED** [review]

**Comment by unguiculus** [comment]
ping @lachie83

**Comment by lachie83** [comment]
I agree with your review @unguiculus

**Comment by mitom** [comment]
I have made the changes you requested and also rebased the branch.

**Comment by unguiculus** [comment]
ping @lachie83

**Review by lachie83 — APPROVED** [review]
lgtm

**Comment by unguiculus** [comment]
@mitom Any chance you can fix the conflicts?

**Comment by pgdagenais** [comment]
@unguiculus conflicts have been fixed in this PR https://github.com/kubernetes/charts/pull/4765 by @timstoop

**Comment by pgdagenais** [comment]
We can now close this PR since https://github.com/kubernetes/charts/pull/4765 have been merged.

**Pull request closed without merging**